src/library/scala/collection/mutable/RedBlackTree.scala

```diff
@@ -0,0 +1,521 @@
+package scala.collection.mutable
+
+import scala.annotation.tailrec
+import scala.collection.Iterator
+
+/**
+ * An object containing the red-black tree implementation used by mutable `TreeMaps`.
+ *
+ * The trees implemented in this object are *not* thread safe.
+ *
+ * @author Rui Gonçalves
+ * @version 2.12
+ * @since 2.12
+ */
+private[collection] object RedBlackTree {
+
+  // ---- class structure ----
+
+  // For performance reasons, this implementation uses `null` references to represent leaves instead of a sentinel node.
+  // Currently, the internal nodes do not store their subtree size - only the tree object keeps track of their size.
+  // Therefore, while obtaining the size of the whole tree is O(1), knowing the number of entries inside a range is O(n)
+  // on the size of the range.
+
+  @SerialVersionUID(21575944040195605L)
+  final class Tree[A, B](var root: Node[A, B], var size: Int) extends Serializable
```

Comment:
Well, I started out by maintaining a `var node: Node[A, B]` in the collection itself, but it would cause  some annoyances in the API of the tree operations.

`insert` and `delete` may cause the root node to change and in the current implementation it is not always easy to keep and return a reference to the new node (the root node can be changed inside `transplant`, for example. Moreover, those operations rely on querying and comparing nodes against the "current" root node, so I would have to have some way of passing a reference to the root node which was always up-to-date. My solution was to create the explicit `Tree` object which despite bringing the extra pointer dereference keeps the code clean and the red-black tree implementation clearly separated from the `TreeMap` implementation (`mutable.TreeSet` could benefit from this implementation too, I believe).